codec/codectools/token.go

```diff
@@ -0,0 +1,78 @@
+package codectools
+
+import (
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+)
+
+type Token struct {
+	Kind TokenKind
+
+	Length int       // Present for MapOpen or ListOpen.  May be -1 for "unknown" (e.g. a json tokenizer will yield this).
```

Comment:
Can someone talk me through the Go memory model decision process around structs like this that are essentially unions. If you make a new `Token` you usually only want to set `Kind` and one of the other properties, yet you have a memory layout that includes space for all of these things. An alternative might be to make everything a pointer, but then you allocate pointer space for each of them and you end up using the heap for the values. Is that the essential trade-off here, and where you have such a choice you should just choose this style? I've run into this a number of times and using pointers always _feels_ more correct to me because I'm not putting anything into the fields I don't care about. But I suppose that's not the best way of thinking about it?

Reply:
Yep, I think you've got it already.

If Go had unions -- either discriminated or even nasty dirty unsafe undiscriminated -- I'd use 'em in a heartbeat.  Unfortunately, it does not.

If one tries to unify the space cost by using any kind of `interface{}`, you end up forcing allocations to box up the values.  (Interfaces have to be two words: one word is a pointer to the type info, and the other word is a pointer to the value.  That means you have to get a pointer from somewhere.  And at present: yes, for _everything_: It's a pointer to the value even if the value is a single word and could just as well fit without a pointer indirection.  Uuffah.  (I've heard talk future versions of Go might improve this situation for word-sized types, but it hasn't happened yet.  And wouldn't entirely absolve our issues here either, since byte slices and strings are both multi-word too.))

("Forcing" allocations might be a little strong: if you can get a pointer to data that's already in the heap (or _inside_ some other data that's already on the heap), then you can get that pointer without a new allocation.  But this generally requires some careful logic... and while sometimes that careful logic is worth the effort, it turns out to be pretty hard to apply in practice around how these tokens get used, IME.)

If one rewrote this structure with the same fields but all pointers (and not trying to unify stuff into a single `interface{}` field), the struct would get about three words shorter (`[]byte` is three words while `*[]byte` is one; `string` is two words while `*string` is one; the rest are all already one word and would still be one word as a pointer; except, okay, `Link` is also still two words and would remain two words because you wouldn't turn it into a pointer-to-an-interface since interfaces can already contain pointers).  So a lot of space would still be reserved.  And you'd indeed still have the same issues with getting a pointer tending to provoke an allocation.

You'll see people still use pointers for optional fields in golang fairly often, but usually the reason for this is that they're wanting to be able to use `nil` as a distinct sentinel value.  (And when you see this, implicitly, the author also either didn't know or just didn't care about the allocations provoked.  Which -- sometimes, is fair!  But in this context, we really do care; this structure is getting used in hot loops in parsers, so allocations are extremely, extremely noticeable.)  A lot of other programming languages try to encapsulate this concept with a type called `Maybe` or `Optional` or `Either` or something like that just to make it clear when the goal is to have a sentinel value -- and we've done it ourselves with the `optional` keyword in IPLD Schemas -- but Golang doesn't have any of these, so, people frequently overload this semantic onto making something into pointer even though it otherwise wouldn't be.

(Sorry for the wall of text; you asked about the one thing that touches all three of my biggest beefs about Golang -- lack of unions, avoidable heap allocations for word-sized residents of interfaces, and the pointer/optional bamboozle -- at once, ahheh.)